### DISCUSSION TITLE
ver1.0-RC1 - v1.0.31106.1

### DISCUSSION THREAD

**Discussion by automatic9045** [automatic9045/BveEX#21] [Announcements]
**本バージョンは RC 版です。正式版リリースまでに少なくとも 1 回、重大な破壊的変更を予定しています。本番環境での採用はお控えください。**

AtsEX についてはこちら→ [AtsEX ATSプラグイン拡張キット - 岡オカ営業所](https://automatic9045.github.io/contents/bve/AtsEX/)

## ⚠️RC 版につき、フィードバックにご協力ください
本バージョンは RC 版です。ひと通りの機能は揃っていますが、動作が不安定な可能性があります。また、仕様を十分に検討し切れていない部分もあります。  
インストールや AtsEX プラグインの使用・開発はもちろん自由に行うことができますが、その際には是非 [Twitter](https://twitter.com/atF9045) や GMail にて **動作報告・不具合報告・仕様に関するご意見** など頂けるととても嬉しいです。

### 更に幅広いデータを操作可能に
BveTypes.dll の機能が更に充実。**車掌、音声、自軌道、補助表示などなど**、AtsEX プラグインから操作できるものの幅が格段にパワーアップしました！

### 更にプラグインを作りやすく
AtsEx.PluginHost、AtsEx.CoreExtensions などの **AtsEX プラグインから参照する DLL を NuGet に公開**。新規にプラグインを開発する際の手間が大幅に低減されました！

その他、車両ファイルが複数ある場合に車両プラグインリストファイル・車両設定ファイルを一括で定義できるようになったりと、プラグインの開発・組込にて地味～に手間がかかっていた部分を細かくじっくりと見直しました。

### 【ご意見募集中】パフォーマンスのための独自ヘッダー・独自ステートメントの仕様変更
従来は何も設定せずとも独自ヘッダー、独自ステートメントはその機能をフルに使用することができましたが、行数の多いマップなどでは読込時間がとても長くなってしまっていました。  
そこで、本バージョンよりマップファイル・AtsEX プラグイン内で特定のフラグをオンにした場合のみ機能を有効化する仕様に変更しました。

読込時間の短縮のために手間が増えた格好ですが、いかがでしょうか。お使いになった際には「こういう仕様にした方が良いのでは？」など是非ご意見をお寄せください。

### 共通
- **【重要】** AtsEx.PluginHost、AtsEx.CoreExtensions を NuGet パッケージとして公開しました。→ 9124a5d8a370f9f2839da23a037f9e40247be6a1
- 多くのコードを整理しました。

### AtsEX 本体
- **【重要】** AtsEX プラグインの Tick メソッドの elapsed パラメータにおいて、ポーズ・早送りが考慮されない不具合を修正しました。→ 5571438445b9aa3e88eabfa911673e5ac3795616
- シナリオが開かれたとき / 閉じられたときをイベントから検知できるようになりました。→ 1fc2aee676a5660c2e09f0eee48dd8330251e240 / e1f25078e33600a25e3824b1cdabb4f2dfe2feee

#### 入力デバイスプラグイン版のみ関係するもの
- **【重要】** ATS プラグインによるブレーキが作用しなくなる不具合を修正しました。→ f046754685be865d091e86b35978e6e9db404b6d
  - この変更は v0.19.30520.1-fix にて適用済です。
- AtsEX 本体の読込でエラーが発生した場合、その詳細が表示されるようになりました。→ 9ec0a6f3ca22920322577915474d7180bb7b66c0
- 終了時にエラーが発生する不具合を修正しました。→ 8cf6109177cc514207e83530a03410218e9e2d7e
- シナリオを再読込するとエラーが発生する不具合を修正しました。→ e2261e13922f1efdc7f9ed8a13bb220cf14d87ba / 573056d3af6c9628274f1f1413f7131bca248106

#### 独自ヘッダー
- **【破壊的変更】** マップファイル冒頭に特定のヘッダーが定義されていない限り、そのマップファイルについてはヘッダーの解釈をスキップする仕様に変更しました。→ 2b4bca7265db06b531c3315503db390007acab89 / b9775a658d8ae5c21db9301c014bd3d747d17e83 / edf74d6a6a28ba1be63af7195a7c04a5a484370e
  - パフォーマンスのためです。従来の仕様では、無条件に全マップファイルの構文解釈を行っていたため、AtsEX を使用している状態で重量級のマップを読み込もうとすると読込に要する時間が極端に長くなってしまっていました。
  - この機能については、現在仕様について検討している段階です。ver1.0 RC → ver1.0 へのバージョンアップ時に仕様が多少変更となる可能性があります。**ご意見等あれば是非お聞かせください。**
- 読込に際して、削除対象でないエラーも削除されてしまう不具合を修正しました。→ fea7c6a67d7693b1cda6c6973d87302b1696a06f / cd575caa3d550b2ab752d088c2fb9cd63a90475b
- シナリオ終了時にメモリリークが発生する不具合を修正しました。→ 06d41e6f614ba1e786d3b909edac76293d83f9ae

#### 独自ステートメント
- **【破壊的変更】** 監視対象の他列車は AtsEX プラグインから指定する仕様に変更しました。→ d5363ae22b9c93d4334a976ca827f12b43854781
  - パフォーマンスのためです。

### プラグインリストファイル
- **【重要】** 同一ディレクトリ内の車両ファイルに対して、1 つのプラグインリストファイルを一括で適用できるようになりました。→ e603d735ffb3dd8a2ca6fd8271a7b282fcc52c79
  - ファイル名を `VehiclePluginUsing.xml` にして車両ファイル群と同一のディレクトリに配置してください。
- **【重要】** プラグインリストファイルのファイル名として `(車両ファイル名).VehiclePluginUsing.xml` 以外の名前も使用できるようになりました。→ e603d735ffb3dd8a2ca6fd8271a7b282fcc52c79
  - 車両設定ファイルに `<PluginIUsingPath>ファイル名</PluginIUsingPath>` と設定することで有効化されます。
- C# アセンブリ / C# スクリプト / IronPython 2 いずれの形態かに関わらず、プラグインの読込順を任意に指定できるようになりました。→ 097bd9a14986b58b8dd4a50381c174075c457777
- ATS プラグイン版でも、AtsEX Caller に付属させる形のみではなく車両ファイルに付属させる形のプラグインリストファイルも読み込めるようになりました。→ e603d735ffb3dd8a2ca6fd8271a7b282fcc52c79

### 車両設定ファイル
- 同一ディレクトリ内の車両ファイルに対して、1 つの車両設定ファイルを一括で適用できるようになりました。→ 7554ee1aaf1fb05e9fa12fc8673394b82bf2505d
  - ファイル名を `VehicleConfig.xml` にして車両ファイル群と同一のディレクトリに配置してください。
- 要素を任意の順番で定義できるようになりました。→ 15182873c60b5aea05dd66d155fffbc50f4079a2
- ATS プラグイン版でも、AtsEX Caller に付属させる形のみではなく車両ファイルに付属させる形の車両設定ファイルも読み込めるようになりました。→ 7554ee1aaf1fb05e9fa12fc8673394b82bf2505d

### AtsEX コア拡張機能
- BVE 内部で定義されている車掌動作を、AtsEX プラグインから上書きできるようにする機能 `ConductorPatch` を追加しました。→ 6c29ec57ecf290629fe54f88ef24c0783da3ab85

### AtsEX 拡張機能: DXDynamicTexture
- テクスチャのファイルパスの表記揺れを許容する仕様に変更しました。→ 068cfbf6f093c6b4c8e099d402f36ccd3bf52c5e

### BveTypes (クラスラッパー除く)
- **【破壊的変更】** LoadAsync メソッドを廃止しました。→ 41fcc327fdf1f2a4b17582aaf292b69861473a7f
  - 同期版と比較して、パフォーマンスなどのアドバンテージに乏しいためです。
- インターフェイスに対応しました。→ 4ab005090b498d79846e3c4ea88b1cd4c55f2da7
- デリゲートに対応しました。→ fd529d3cd2e986b0457be4dc9c6ad485e06bf232

### クラスラッパー
- **【破壊的変更】** StructureBlock.MatrixFromLastBlock プロパティを TransformFromLastBlock に改名しました。→ 5e52a79fd6ff5aa35b3c23d92657afc5987b59f2
- **【破壊的変更】** AssistantTextBase クラス、AssistantTextDraw クラスは、命名が不適切だったため改名しました。→ f0e2fc340f61734f187bc979a585979342343a37
- Direct3DProvider クラスから DirectSound オブジェクトを取得できるようになりました。→ 0124454aa82b99a3096ea9970e9c32a7791f7877
- 車掌関連のクラス (Conductor クラス、Passenger クラス) を追加しました。→ 8dd0bef163cdfda6944474d18f5cba358b951ea5 / b72eed449aa364d75a2a8c547092f338997baee5 / df0c28c3a136e9a214a07d03e141ab2c39488e49 / 8924d336345c3861bc3bf0e175d64b4b2f86c903
- SoundSet クラスを追加しました。→ 8dd0bef163cdfda6944474d18f5cba358b951ea5
- Sound クラスのコンストラクタを追加しました。→ ee66e5be98aecbffa73838ccbbbf59592c1bc471
- Sound クラスから音声を再生できるようになりました。→ 8dd0bef163cdfda6944474d18f5cba358b951ea5 / 8906c63f8d68656fc7e672c8539d81f8a3c80658
- Sound クラスから音声の設置位置、速度を操作できるようになりました。→ 4737b376000f30b4ceed5724add72d448a1122ae
  - これにより、いわゆる Sound3D でループ以外の任意の音声も再生できるようになりました。
- SectionManager クラスから前後の閉そくの制限速度を取得できるようになりました。
- 自軌道関連のクラス (Cant クラス、CantList クラス、Curve クラス、CurveList クラス、MyTrack クラス) を追加しました。→ 0398861193b6ef80901731e67dd3826379620154 / c62f2b00ea423480b342aed010d8404ddf15d5ab
  - 自軌道の曲線、カントの位置及び量を全て取得できるようになりました。
- 自列車の位置を制御するためのクラス (Transform クラス、VehicleBogieWheel クラス、VehiclePositioner クラス、VehicleVibrationManager クラス) を追加しました。→ 
  - 車体を傾けたり、自列車を線路から離れて動かすことが可能となります。
- 視点を表す CameraLocation クラスを追加しました。→ 15c602c6cf800abfb81ab8352f70c75df636ff57 / 1c1b05769983ecea77d165960130c9a96a29e8c3
  - 視点を編集できるようになりました。
- 自列車のバネに関するクラス (AirSpring クラス、Spring クラス) を追加しました。→ b1d3ed7701cfed2116570d49da3297ac94ef4fe5
- 6DoFを表す SixDof クラスを追加しました。→ b1d3ed7701cfed2116570d49da3297ac94ef4fe5
- 背景モデルを表す Background クラスを追加しました。→ 8d78997145cd1d0bbf6cdd50d534eec3601deb7f
- 補助表示関連のクラス (AssistantBase クラス、AssistantDrawer クラス、ClockText クラス、DetailText クラス、StationText クラス、SpeedLimitText クラス、GeometryText クラス、LeverText クラス、InstructionText クラス、SwayGauge クラス、StopPositionGauge クラス) を追加しました。→ f0e2fc340f61734f187bc979a585979342343a37
- InterpolatableMapObjectList クラスを追加しました。→ 7bebd1336c50e03b1afdc8f71de22f17884a2e03
- IDrawable インターフェイスを追加し、AssistantBase クラス、AssistantDrawer クラス、MainForm クラスにその実装を追加しました。→ 1c1b05769983ecea77d165960130c9a96a29e8c3
- ReAdhesionControl クラス、ReAdhesionControlMode 列挙型を追加しました。→ 4a8bf142002bcd9de8f71e30f69ff5b9d280af34
  - 車両の再粘着制御の動作状態を取得できるようになりました。
- Scenario.Dispose メソッドを追加しました。→ cbe3f0cf6459cc99a5774e565356c3492af67f2f
- MapObjectList.GoTo メソッド、MapObjectList.GoToAndGetCurrent メソッドを追加しました。→ a5e28519223355e128305b16814a2114dcbc9c67 / 9edc4239b7e128209bfe9486f7a1ec92e0710c9d
  - 現在の距離程に対応する要素を簡単に取得できるようになりました。
- Station.DoorSideEnum プロパティを追加しました。→ 861d1b26d35eaba8efcb07369a8bde9e603765c5
  - Conductor との連携を容易にするためです。
- StructureBlock.Direction プロパティを追加しました。→ b088bd7732acc062e78fa6f3e65f1b78b60a9fa2
  - ストラクチャーブロックの方角が分かるため、自列車を線路から離れて動かしたいときなどに便利です。
- DrawDistanceManager クラスから描画距離を操作できるようになりました。→ 30ffb8fb1d18f1cd212354f1f1acacefbf71a316
- Route.CabIlluminanceObjects プロパティを追加しました。→ 7bebd1336c50e03b1afdc8f71de22f17884a2e03
  - 運転台の明るさを操作できるようになりました。
- Direct3DProvider クラス、ObjectDrawer クラス、Scenario クラスへ、オブジェクト類を描画するときに呼び出されるメソッドを追加しました。→ 1c1b05769983ecea77d165960130c9a96a29e8c3
- Train、TrainInfo クラスのコンストラクタを追加しました。→ 53dff43a028cf94a52a085b96eb954258852b901
- MapObjectList.ObjectPassed イベントを追加しました。→ 63b909fdd51ebfa7a7da1d994e80334d383c3432
  - マップオブジェクトの通過時に通知します。

### FastMember
- イベントに対応しました。→ f5931cc591b9900054403d59e6614ccc8a03e7e4 / e07900f9260c3ef9daf298a3a5cdd2ee65fccf78

### TypeWrapping
- イベントに対応しました。→ 2f59b87ee6634650574d3724a8665d776cc2fbe7
- デリゲートに対応しました。→ fd529d3cd2e986b0457be4dc9c6ad485e06bf232

### ObjectiveHarmonyPatch
- **【重要】** NuGet パッケージとして別リポジトリに分離しました。→ 8463a0e55d7667e8b33bdf76565357b18a021961

### UnembeddedResources
- **【重要】** NuGet パッケージとして別リポジトリに分離しました。→ da683f88e428df778fdf75bf6c6eb141e185d2fa
- AtsEX プラグイン等の開発時、XML ドキュメントを表示できるようになりました。→ d241bc75bcb9e63704cbb03b8ee32b13218876bb

### サンプル
- `ConductorPatch` 拡張機能のサンプルを追加しました。→ 754cfd8e71a6790be3a7cebb9e541d4c78381597
- E217 系車両データを同梱しました。→ fa52ba28c0410114717d2d3f368796b1ae4e726e / 2e8bae63cb7c753f8f02c55df1616d64f5328e13 / e5d9d6f0debaeb84d5f29b6b0daa0ef8c499903e
- 各シナリオにサムネイルを追加しました。→ 6881a983d0fc508e21aafbcecbedcf9205fd7770
- AtsEX 非搭載時に表示される NOMPI メッセージを、現在の仕様に合った内容に変更しました。→ 3e2360ed44a29243b97abe6015f593f910c9ffba

### 拡張機能
- AtsEX コア拡張機能
  - **【新規】** ConductorPatch
    - 車掌の動作を自由に変更できるようにするパッチを提供します。
  - ContextMenuHacker
    - メインフォームの右クリックメニューを編集するための機能を提供します。
  - DiagramUpdater
    - 時刻表、ダイヤグラムなどの行路に関わるオブジェクトの更新機能を提供します。
  - PreTrainPatch
    - 先行列車の走行位置を自由に変更できるようにするパッチを提供します。
  - SignalPatch
    - 閉塞の信号現示を自由に変更できるようにするパッチを提供します。
  - StructureModelFinder
    - ストラクチャーの 3D モデルを簡単に検索するための機能を提供します。
  - TrainDrawPatch
    - 他列車を自由に移動・回転できるようにするパッチを提供します。
- DXDynamicTexture
- \* 拡張地上子
  - 古い機能のため非推奨。独自マップ構文機能を使用してください。

### 車両プラグイン
- 運転台パネルに値を渡すサンプル
- **【新規】** 車掌サンプル
- マッププラグイン連携サンプル
- シンプルな ATS サンプル
- 状態取得・設定サンプル

### マッププラグイン
- 車両プラグイン連携サンプル
- DXDynamicTexture 連携サンプル
- 独自マップ構文サンプル
- 先行列車操作サンプル
- 信号現示操作サンプル
- 駅リスト編集サンプル
- 他列車操作サンプル
- 他列車操作サンプル EX
- 走行抵抗係数編集サンプル
- 空の C# スクリプトプラグイン
- 空の IronPython プラグイン

### AtsEX 本体＋サンプルシナリオ 
AtsEX 本体に加えて、車両プラグイン・マッププラグインの動作が確認できるサンプルシナリオを同梱しています。

#### 【推奨】入力デバイスプラグイン・インストーラー版
下の Assets より「AtsEX_v1.0.31106.1.Setup.exe」をダウンロードし、実行してください。

⚠️**実行には管理者権限が必要です。**  
ご家族内で共有の PC を使用している場合などで、管理者権限を得られないときは、下の「入力デバイスプラグイン・手動インストール版」をご利用ください。

#### 入力デバイスプラグイン・手動インストール版
下の Assets より「AtsEX_v1.0.31106.1_AsInputDevice.7z」をダウンロード・解凍し、シナリオフォルダおよび BVE 本体の Input Devices フォルダに配置してください。

インストール方法の説明はこちら→ [AtsEXインストールガイド - AtsEX ATSプラグイン拡張キット - 岡オカ営業所](https://automatic9045.github.io/contents/bve/AtsEX/docs/user/installation-guide.html)

#### ATS プラグイン版
下の Assets より「AtsEX_v1.0.31106.1_AsAtsPlugin.7z」をダウンロード・解凍し、シナリオフォルダに配置してください。

### AtsEX SDK
サンプルプラグインのソースコードのセットです。

下の Assets より「AtsEX_v1.0.31106.1_SDK.7z」をダウンロード・解凍してください。  
ビルドには別途 AtsEX 本体のダウンロードが必要です。

解凍には [7-zip](https://sevenzip.osdn.jp/) などの解凍ソフトが必要です。

<hr /><em>This discussion was created from the release <a href='https://github.com/automatic9045/AtsEX/releases/tag/v1.0.31106.1'>ver1.0-RC1 - v1.0.31106.1</a>.</em>

**Comment by winup-zhou** [comment]
Visual StudioのNuGet Package Managerからこのバージョンを取得できません。
どうすれば更新できますか？

**Reply by automatic9045** [reply]
NuGet の [プレリリースを含む] チェックボックスにチェックを付けると、ver1.0.0-rc1 が選択肢に表示されるようになります。
